presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDirectStreamReader.java

```diff
@@ -150,41 +155,61 @@ public Block readBlock(Type type)
             throw new PrestoException(GENERIC_INTERNAL_ERROR,
                     format("Values in column \"%s\" are too large to process for Presto. %s column values are larger than 1GB [%s]", streamDescriptor.getFieldName(), nextBatchSize, streamDescriptor.getOrcDataSourceId()));
         }
-
-        byte[] data = EMPTY_BYTE_ARRAY;
-        if (totalLength > 0) {
-            if (dataStream == null) {
-                throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
-            }
-            data = dataStream.next(toIntExact(totalLength));
+        if (dataStream == null) {
+            throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
         }
 
-        Slice[] sliceVector = new Slice[nextBatchSize];
-
-        int offset = 0;
-        for (int i = 0; i < nextBatchSize; i++) {
-            if (!isNullVector[i]) {
-                int length = lengthVector[i];
-                Slice value = Slices.wrappedBuffer(data, offset, length);
-                if (isVarcharType(type)) {
-                    value = truncateToLength(value, type);
-                }
-                if (isCharType(type)) {
-                    value = trimSpacesAndTruncateToLength(value, type);
-                }
-                sliceVector[i] = value;
-                offset += length;
+        // allocate enough space to read
+        byte[] data = new byte[toIntExact(totalLength)];
+        Slice slice = Slices.wrappedBuffer(data);
+
+        // truncate string and update offsets
+        offsetVector[0] = 0;
+        for (int i = 0; i < currentBatchSize; i++) {
+            // the previously computed offsets are original offsets without truncation
+            // update the offsets in this loop to be the truncated ones
+            if (isNullVector[i]) {
+                offsetVector[i + 1] = offsetVector[i];
+                continue;
+            }
+            int offset = offsetVector[i];
+            int length = lengthVector[i];
+
+            // read data without truncation
+            dataStream.next(offset + length, data, offset);
+
+            // calculate truncated length
+            int truncatedLength = length;
+            if (isVarcharType(type)) {
+                VarcharType varcharType = (VarcharType) type;
+                int codePointCount = varcharType.isUnbounded() ? length : varcharType.getLengthSafe();
+                truncatedLength = getTruncatedLength(slice, offset, length, codePointCount);
+            }
+            else if (isCharType(type)) {
+                int trimmedLength = sliceLengthWithoutTrailingSpaces(slice, offset, length);
+                verify(trimmedLength >= 0);
+                int codePointCount = ((CharType) type).getLength();
+                truncatedLength = getTruncatedLength(slice, offset, trimmedLength, codePointCount);
```

Comment:
Trimming before truncation can lead to invalid `char`s. Do it the other way around.

Turn trim and truncate into one function. And test that function.